Add support for Model Context Protocol

martin-fleck-at opened the pull request:
#### What it does

MCP client-side support paired with the server-side integration, plus
a generic export pipeline that underpins server-driven export flows
and is reusable by UI flows.

MCP integration
- Adds an opt-in MCP server configuration on initialize and a
  server descriptor on the result.
- Splits the configuration into init- and deploy-controllable parts
  so security-relevant fields stay under deployment control.
- Enable MCP by setting the configuration on initialize; the
  standalone example demonstrates the wiring.

Generic export pipeline
- Format-agnostic export request/response pair, coexisting with the
  legacy SVG-only flow under strict separation.
- Pluggable export-strategy registry with shared post-processing.
- Ships default SVG and PNG strategies; PNG export is a new
  built-in capability.
- Register a custom strategy to add formats or override defaults.

Misc
- Promotes the origin-viewport action to the protocol so the server
  can dispatch it.
- Widens the shutdown contract to allow asynchronous handlers.

Part of https://github.com/eclipse-glsp/glsp/issues/1546

#### How to test

- Nothing breaks
- For the overall change, see [GLSP Node Server](https://github.com/eclipse-glsp/glsp-server-node/pull/120) and [Theia Integration](https://github.com/eclipse-glsp/glsp-theia-integration/pull/271)

#### Follow-ups

<!-- Please list potential follow-up work, including known issues, possible future work, identified technical debt, and potentially introduced technical debt. If the PR introduces technical debt, specify the reason why this is acceptable. Please create tickets and link them here. Please use the label "technical debt" for new issues when it applies. -->

#### Changelog

<!-- Please check, when if it applies to your change. -->

-   [x] This PR should be mentioned in the changelog
-   [ ] This PR introduces a breaking change (if yes, provide more details below for the changelog and the migration guide)

tortmayr reviewed (changes requested):
In general the changes look good to me.
I only have a fewer minor nitpicks.

martin-fleck-at committed:
feat(mcp): add Model Context Protocol client-side support

MCP client-side support paired with the server-side integration, plus
a generic export pipeline that underpins server-driven export flows
and is reusable by UI flows.

MCP integration
- Adds an opt-in MCP server configuration on initialize and a
  server descriptor on the result.
- Splits the configuration into init- and deploy-controllable parts
  so security-relevant fields stay under deployment control.
- Enable MCP by setting the configuration on initialize; the
  standalone example demonstrates the wiring.

Generic export pipeline
- Format-agnostic export request/response pair, coexisting with the
  legacy SVG-only flow under strict separation.
- Pluggable export-strategy registry with shared post-processing.
- Ships default SVG and PNG strategies; PNG export is a new
  built-in capability.
- Register a custom strategy to add formats or override defaults.

Misc
- Promotes the origin-viewport action to the protocol so the server
  can dispatch it.
- Widens the shutdown contract to allow asynchronous handlers.

Part of https://github.com/eclipse-glsp/glsp/issues/1546

Co-authored-by: Andreas Hell <44035624+Sakrafux@users.noreply.github.com>

martin-fleck-at commented on packages/protocol/src/client-server-protocol/mcp.ts:
```diff
@@ -0,0 +1,78 @@
+    mcpServer: McpServerConfiguration;
+}
+
+export function isMcpInitializeParameters(params?: InitializeParameters): params is McpInitializeParameters {
```
Done.

martin-fleck-at commented on packages/protocol/src/client-server-protocol/mcp.ts:
```diff
@@ -0,0 +1,78 @@
+}
+
+export function isMcpInitializeParameters(params?: InitializeParameters): params is McpInitializeParameters {
+    return !!params && (params as McpInitializeParameters).mcpServer !== undefined;
```
Done.

martin-fleck-at commented on packages/protocol/src/client-server-protocol/mcp.ts:
```diff
@@ -0,0 +1,78 @@
+    mcpServer: McpServerResult;
+}
+
+export function isMcpInitializeResult(result?: InitializeResult): result is McpInitializeResult {
```
Done.

tortmayr commented on examples/workflow-standalone/src/node/app.ts:
```diff
@@ -47,7 +50,22 @@ async function initialize(connectionProvider: MessageConnection, isReconnecting
+
+    // Surface the MCP server URL the GLSP server announced in its `InitializeResult`. The IDE
+    // integrations consume this programmatically; standalone has no native consumer, so log it
+    // and surface a toast for dev visibility. `MessageAction` falls through to the standalone's
```
Lots of comment fluff.
The fist two sentences should be more than enough.
I don't think the extended reasoning adds much value

martin-fleck-at replied:
I'll reduce it.

tortmayr commented on packages/client/src/default-modules.ts:
```diff
@@ -1,5 +1,5 @@
 /********************************************************************************
- * Copyright (c) 2021-2025 EclipseSource and others.
+ * Copyright (c) 2021-2026 EclipseSource and others.
```
Seems like an accidental change. The file was last touched in 2025.

martin-fleck-at replied:
I had to amend this commit as otherwise it will always be tagged as touched by the CI. All other issues are resolved through a follow-up commit.

tortmayr commented on packages/protocol/src/action-protocol/viewport.ts:
```diff
@@ -111,6 +111,38 @@ export namespace FitToScreenAction {
+ * The corresponding namespace declares the action kind as constant and offers helper functions for type guard checks
+ * and creating new `OriginViewportActions`.
+ */
+export interface OriginViewportAction extends Action {
```
Seems like we are missing tests for the new action in viewport.spec.ts

martin-fleck-at replied:
Added.

tortmayr commented on CHANGELOG.md:
```diff
@@ -7,10 +7,20 @@
 ### Potentially Breaking Changes
 
 -   [node] Update minimum requirements for Node to >=22 [#476](https://github.com/eclipse-glsp/glsp-client/pull/476)
+-   [protocol] `OriginViewportAction.is` now also requires the `animate` field (matching the shape of `CenterAction.is` and `FitToScreenAction.is`). Hand-rolled emitters that ship a bare `{kind: 'originViewport'}` will no longer pass the guard; use `OriginViewportAction.create()` (which defaults `animate: true`) or include `animate` explicitly. [#456](https://github.com/eclipse-glsp/glsp-client/p ...
```
Im confused. What do you mean with that?
The `OriginViewportAction` on master already had a required `animate` field. So What is the break here?

martin-fleck-at replied:
It did have it but the type check didn't evaluate it, i.e., it was:
```
export namespace OriginViewportAction {
    export function is(object: any): object is OriginViewportAction {
        return Action.hasKind(object, KIND);
    }
}
```

But the new type check does check whether animate is actually set. If the action was created through our `create` function, it should not make a difference, just wanted to note it down.

tortmayr replied:
Ah I see.  It's just for the type check. yeah okay that's fine.

tortmayr commented on examples/workflow-standalone/src/node/app.ts:
```diff
@@ -47,7 +50,22 @@ async function initialize(connectionProvider: MessageConnection, isReconnecting
+    // and surface a toast for dev visibility. `MessageAction` falls through to the standalone's
+    // FallbackActionHandler (logs only); `ShowToastMessageAction` (from the accessibility module
+    // which is loaded in this standalone) renders a real UI toast that persists until dismissed.
+    const mcpServer = McpInitializeResult.getServer(glspClient.initializeResult);
```
Maybe we should use a fixed port for the mcp server in the standalone case?
Just to make testing easier

martin-fleck-at replied:
That's a good case! I will follow what we do for the GLSP Port with the environment variable, I think that's a good approach, agree?

tortmayr replied:
Yeah sounds good,  choose a default port and make it configurable via env. similar to how it is done for the glsp sever port itself

martin-fleck-at replied:
Added.

tortmayr commented on packages/client/src/features/export/export-modules.ts:
```diff
@@ -17,22 +17,38 @@ import {
-        bind(TYPES.SvgExporter).to(GLSPSvgExporter).inSingletonScope();
+        bindAsService(context, TYPES.SvgExporter, GLSPSvgExporter);
+
+        // Unified export path — `RequestExportAction` / `ExportResultAction` + DiagramExporter registry.
```
Remove this generated comment please.
It explains a common pattern that is used across the entire glsp code base.
We should not start to document this in every module etc.

martin-fleck-at replied:
Done.

tortmayr commented on packages/protocol/src/action-protocol/model-saving.ts:
```diff
@@ -143,3 +143,169 @@ export namespace ExportSvgAction {
+/**
+ * Format identifier for a diagram export. The package ships `'svg'` and `'png'`;
+ * adopters add formats by registering a `DiagramExporter` strategy keyed on any string.
+ * `ProposalString` gives autocomplete for the shipped values without closing the union —
```
Keep it concise we don't need to explain the purpose of propose string on every use.

martin-fleck-at replied:
Done

tortmayr commented on packages/protocol/src/action-protocol/model-saving.ts:
```diff
@@ -143,3 +143,169 @@ export namespace ExportSvgAction {
+ */
+export type ExportFormat = ProposalString<'svg' | 'png'>;
+
+/** MIME type returned alongside an export. `ProposalString` for the same reason as {@link ExportFormat}. */
```
```suggestion
/** MIME type returned alongside an export. */
```

martin-fleck-at replied:
Done

tortmayr commented on packages/protocol/src/action-protocol/model-saving.ts:
```diff
@@ -143,3 +143,169 @@ export namespace ExportSvgAction {
+ * Options shared by every export strategy that renders through sprotty's SVG export pipeline.
+ * Today honoured by SVG output and the canvas-based PNG rasteriser; future strategies that
+ * also raster from SVG (PDF, PPTX, …) would extend this interface for the same knob. Lives at
+ * the protocol level so server-side dispatchers (e.g. MCP resource handlers) can populate the
```
Drop last line

martin-fleck-at replied:
Done

tortmayr commented on packages/protocol/src/action-protocol/model-saving.ts:
```diff
@@ -143,3 +143,169 @@ export namespace ExportSvgAction {
+    /**
+     * Skip the per-element style-copy step. The default copy walk is expensive on large
+     * diagrams; setting this flag skips it at the cost of losing CSS-driven styling in the
+     * exported output. Equivalent to sprotty's `ExportSvgOptions.skipCopyStyles`.
```
```suggestion
     * exported output.
```

martin-fleck-at replied:
Done

tortmayr commented on packages/protocol/src/action-protocol/model-saving.ts:
```diff
@@ -143,3 +143,169 @@ export namespace ExportSvgAction {
+    /** General overload for adopter-registered formats; `formatOptions` is opaque. */
+    export function create(format: ExportFormat, options?: { formatOptions?: unknown; requestId?: string }): RequestExportAction;
+    export function create(format: ExportFormat, options: { formatOptions?: unknown; requestId?: string } = {}): RequestExportAction {
+        // Generated id by default (mirrors `RequestExportSvgAction.create`) so UI-direct
```
Remove. This is the default behavior of every request action. not warrants a comment

martin-fleck-at replied:
Is it? Most of our request actions initialize with `requestId: ''` and the `requestId` is then set if you use our `request` or `requestUntil` method of the action dispatcher. So it is not exactly the same and we anted to match the `RequestExportSvgAction` behavior. But I just saw that we override that with our empty string anyways, only the sprotty version of it, set the id directly. I'll use the empty string.

tortmayr commented on packages/client/src/features/export/default-png-diagram-exporter.ts:
```diff
@@ -0,0 +1,139 @@
+
+/**
+ * Default fallback width (in CSS px) when the request omits an explicit `width`.
+ * Matches the previous MCP-specific PNG path so existing MCP consumers see no
```
Matches what preexisting behavior?
Comment seems of.

martin-fleck-at replied:
Good catch, will fix.

tortmayr commented on packages/client/src/features/export/default-png-diagram-exporter.ts:
```diff
@@ -0,0 +1,139 @@
+/**
+ * Default PNG strategy for the unified export registry. Reuses {@link GLSPSvgExporter}'s
+ * SVG-rendering pipeline and rasterises the result client-side via `OffscreenCanvas` —
+ * browser-native, no extra dependencies. Adopters that need server-side rendering (e.g.
```
drop the em-dash additions.

martin-fleck-at replied:
Done

tortmayr commented on packages/client/src/features/export/default-png-diagram-exporter.ts:
```diff
@@ -0,0 +1,139 @@
+        const svgString = this.svgExporter.exportToString(root, options, cause);
+        const { width, height } = await this.computeDimensions(svgString, options);
+        // Pin the SVG root to the target raster size BEFORE loading into `<img>` so the browser
+        // rasterises the vector content at full resolution. Without this the browser uses the
```
First line should be enough for context.

martin-fleck-at replied:
Done

tortmayr commented on packages/client/src/features/export/diagram-export-postprocessor.ts:
```diff
@@ -0,0 +1,86 @@
+import { DiagramExporter } from './diagram-exporter';
+
+/**
+ * Postprocessor for the unified `RequestExportAction` flow. Captures the rendered model
```
```suggestion
 * Postprocessor for the unified {@link RequestExportAction} flow. Captures the rendered model
```

martin-fleck-at replied:
Done

tortmayr commented on packages/client/src/features/export/diagram-export-postprocessor.ts:
```diff
@@ -0,0 +1,86 @@
+
+/**
+ * Postprocessor for the unified `RequestExportAction` flow. Captures the rendered model
+ * root from the `decorate` pass, then on `postUpdate` looks up the matching
```
You can use proper links for the method names (decorate and postupdate)

martin-fleck-at replied:
Done

tortmayr commented on packages/client/src/features/export/diagram-exporter.ts:
```diff
@@ -0,0 +1,45 @@
+import { Action, ExportEncoding, ExportFormat, ExportMimeType, GModelRoot } from '@eclipse-glsp/sprotty';
+
+/**
+ * A pluggable strategy that turns the rendered diagram into the bytes of a specific export
```
Lots of fluff. maybe make it more concise

martin-fleck-at replied:
Done

tortmayr commented on packages/client/src/features/export/export-result-action-handler.ts:
```diff
@@ -0,0 +1,71 @@
+import { injectable } from 'inversify';
+
+/**
+ * Default handler for the unified {@link ExportResultAction}. Triggers a browser file
```
Double check with the default `ExportSvgActionHandler`.
Maybe more concise

martin-fleck-at replied:
Done

tortmayr commented on packages/client/src/features/export/request-export-command.ts:
```diff
@@ -0,0 +1,65 @@
+import { inject } from 'inversify';
+
+/**
+ * `HiddenCommand` for the unified `RequestExportAction` kind. Performs the same
```
Can use proper links here since both Hidden Command and RequestExportAction are imported

martin-fleck-at replied:
Done

tortmayr commented on packages/protocol/src/client-server-protocol/mcp.ts:
```diff
@@ -0,0 +1,270 @@
+ * over the wire, while security-sensitive bind/policy fields stay adopter-controlled and
+ * cannot be widened by an MCP `initialize` payload.
+ */
+/** Wire-protocol naming for the data-handler exposure mode (see {@link McpServerInitOptions.dataMode}). */
```
Dangling jsdoc. should probably be combined in one block

martin-fleck-at replied:
Done

tortmayr commented on packages/client/src/features/export/glsp-svg-exporter.ts:
```diff
@@ -28,17 +29,39 @@ import { v4 as uuid } from 'uuid';
+        if (!svgElement) {
+            throw new Error('SVG export failed: SVG element not found');
+        }
+        const request = cause as RequestExportSvgAction | undefined;
```
When called from DefaultSvgDiagramExporter or DefaultPngDiagramExporter, the cause is a RequestExportAction, not a RequestExportSvgAction. The cast is used to pass to prepareSvgElement and getSvgExport, which take RequestAction<ExportSvgAction>. If an adopter override of those methods does something with request.options (which exists on RequestExportSvgAction but not RequestExportAction), it'll get undefined without a type error. This is probably fine in practice (the base implementations only access requestId), but the cast is technically unsound. Consider documenting this or narrowing what those internal methods actually need.

martin-fleck-at replied:
Added a comment to explain the desired flow.

tortmayr commented on packages/protocol/src/client-server-protocol/mcp.ts:
```diff
@@ -0,0 +1,270 @@
+     * need both the type-cast and the runtime guarantee that the field is present.
+     */
+    export function is(result?: InitializeResult): result is McpInitializeResult & { mcpServer: McpServerResult } {
+        return AnyObject.is(result) && hasObjectProp(result, 'mcpServer');
```
Should we do a deeper check here? `McpServerResult.is(result.mcpServer)`

martin-fleck-at replied:
Added.

tortmayr commented on packages/client/src/features/export/default-png-diagram-exporter.ts:
```diff
@@ -0,0 +1,139 @@
+ * resvg, headless Chrome) replace this binding with their own strategy.
+ */
+@injectable()
+export class DefaultPngDiagramExporter implements DiagramExporter<PngExportOptions> {
```
Looks like the svg image is loaded twice.

computeDimensions creates a bitmap from the original SVG to measure aspect ratio, then rasterise creates another bitmap from the resized SVG. When only width is provided, this results in two full SVG->bitmap conversions. Consider extracting the initial bitmap's dimensions more cheaply (e.g., parsing the SVG viewBox attribute) to avoid the double decode.

martin-fleck-at replied:
Will do, thank you, that is a great point!

tortmayr commented on packages/protocol/src/client-server-protocol/mcp.ts:
```diff
@@ -0,0 +1,270 @@
+    name: string;
+
+    /** The transport type used to talk to the MCP server. */
+    type: 'http';
```
What is the point of a hardcoded type property?

If `http` is the only allowed type its not needed.
If it should be generic it should be a ProposalString

martin-fleck-at replied:
I think we can actually remove it for now. The [spec](https://modelcontextprotocol.io/specification/2025-11-25/basic/transports) defines http or stdio as two possible transport mechanisms. However, introducing this field later when we actually have another to support is easy and now this leads more to confusion, I'll remove it.

martin-fleck-at committed:
fixup: PR feedback

Addresses review on PR #456.

Protocol
- Drop `McpServerResult.type`; require `McpInitializeResult.mcpServer`.
- Tighten `McpServerResult.is` / `McpInitializeResult.is` via GLSP
  type utilities.
- Rename `GLSPServerInitContribution` → `GLSPServerInitializer`.
- Deprecate `RequestExportSvgAction` / `ExportSvgAction`.
- Restore `requestId: ''` on `RequestExportAction.create`.

Export pipeline
- Replace unsound `cause` cast in `GLSPSvgExporter` with a type
  guard.
- `DefaultPngDiagramExporter`: single SVG decode (read aspect from
  attributes).
- New `RequestExportKeyListener` (Ctrl+Shift+E) for the unified
  action; legacy bindings stay wired.
- Deprecate `ExportSvgActionHandler` and `GLSPSvgExporter.export()`.

Misc
- `GLSP_MCP_SERVER_PORT` env-var (default `64577`).
- `OriginViewportAction.is` / `.create` tests.
- JSDoc cleanup: trim verbose comments, linkify backticked refs.

martin-fleck-at reviewed (commented):
Thank you very much, Tobias! I updated the PR and all your concerns should be addressed. One thing to note though is that we are currently keeping the previous ExportSvg bindings to support the old flow but every part of it has been marked as deprecated.

tortmayr reviewed (approved):
LGTM! 👍🏼
